src/rendering.rs

```diff
@@ -0,0 +1,182 @@
+        // XXX The display could be obtained from any object created by it, so we can
+        // query it from the config.
+        let gl_display = gl_config.display();
+        // The context creation part.
+        let context_attributes = ContextAttributesBuilder::new().build(raw_window_handle);
+        // Since glutin by default tries to create OpenGL core context, which may not be
+        // present we should try gles.
+        let fallback_context_attributes = ContextAttributesBuilder::new()
+            .with_context_api(ContextApi::Gles(None))
+            .build(raw_window_handle);
+        // There are also some old devices that support neither modern OpenGL nor GLES.
+        // To support these we can try and create a 2.1 context.
+        let legacy_context_attributes = ContextAttributesBuilder::new()
+            .with_context_api(ContextApi::OpenGl(Some(Version::new(2, 1))))
+            .build(raw_window_handle);
+        let not_current_gl_context = unsafe {
+            gl_display
+                .create_context(&gl_config, &context_attributes)
+                .unwrap_or_else(|_| {
+                    gl_display
+                        .create_context(&gl_config, &fallback_context_attributes)
+                        .unwrap_or_else(|_| {
+                            gl_display
+                                .create_context(&gl_config, &legacy_context_attributes)
+                                .expect("failed to create context")
+                        })
+                })
+        };
+
+        // Create surface
+        let attrs = window
+            .build_surface_attributes(Default::default())
+            .expect("Failed to build surface attributes");
+        let surface = unsafe {
+            gl_config
+                .display()
+                .create_window_surface(&gl_config, &attrs)
+                .unwrap()
+        };
+
+        // Make it current.
+        let context = not_current_gl_context.make_current(&surface).unwrap();
+
+        // Try setting vsync.
+        if let Err(res) =
+            surface.set_swap_interval(&context, SwapInterval::Wait(NonZeroU32::new(1).unwrap()))
+        {
+            log::error!("Error setting vsync: {res:?}");
+        }
+
+        let gl = match context.context_api() {
+            ContextApi::OpenGl(_) => unsafe {
+                gleam::gl::GlFns::load_with(|symbol| {
+                    let symbol = CString::new(symbol).unwrap();
+                    gl_display.get_proc_address(symbol.as_c_str()) as *const _
+                })
+            },
+            ContextApi::Gles(_) => unsafe {
+                gleam::gl::GlesFns::load_with(|symbol| {
+                    let symbol = CString::new(symbol).unwrap();
+                    gl_display.get_proc_address(symbol.as_c_str()) as *const _
+                })
+            },
+        };
+
+        log::debug!("Running on {}", gl.get_string(gl::RENDERER));
```

Comment:
Maybe println is better, so we can always understand our GL context?

Reply:
Agreed, it might be better to just print it out since all of we're quite early in development and GL is pretty prone to error.